In [1]:
%load_ext tensorboard
import tensorflow as tf
import datetime
# Clear any logs from previous runs
#rm -rf ./logs/

In [2]:
# Mengambil data CIFAR-10
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()

170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 95s 1us/step


In [3]:
#Fungsi untuk membuat model
def create_model():
    return tf.keras.models.Sequential([
        tf.keras.layers.Flatten(input_shape=(28, 28)),
        tf.keras.layers.Dense(512, activation='relu'),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Dense(10, activation='softmax')
  ])

# Menyimpan file setelah run tensorboard

In [4]:
log_dir = "new_logs"
tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)

# Membuat model

In [18]:
from tensorflow.keras import layers
from tensorflow import keras

# Model Sequential dengan input layer untuk CIFAR-10
model = keras.Sequential([
    layers.InputLayer(input_shape=(32, 32, 3)),
    # Lapisan lainnya...
])

In [19]:
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
# 64: Jumlah filter atau kernel konvolusi yang digunakan. Semakin banyak filter, semakin banyak fitur yang dapat diekstraksi.
# Ukuran kernel konvolusi, yaitu 3x3. Ini menunjukkan bahwa setiap kernel memiliki ukuran 3x3 piksel.
# activation='relu': Fungsi aktivasi ReLU (Rectified Linear Unit), yang mempertahankan nilai positif dan mengganti nilai negatif dengan nol. 

model.add(layers.MaxPooling2D((2, 2))) #Max pooling memilih nilai maksimum dalam area tertentu (dalam hal ini, 2x2).
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.Flatten()) #Mengubah (flatten) data dari format 2D atau multi-dimensi menjadi satu dimensi. 
model.add(layers.Dense(64, activation='relu')) # Layer fully connected dengan 64 neuron. Setiap neuron terhubung dengan semua neuron pada layer sebelumnya. 
model.add(layers.Dense(10, activation='softmax'))  # Lapisan output untuk klasifikasi 10 kelas


In [13]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',  # Ganti dengan 'categorical_crossentropy' jika label one-hot
              metrics=['accuracy'])  # Ukuran evaluasi


In [14]:
# Misalkan Anda sudah memiliki data CIFAR-10 dengan label yang sesuai
model.fit(x_train, y_train, epochs=10, validation_data=(x_test, y_test))

Epoch 1/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 46s 28ms/step - accuracy: 0.2018 - loss: 4.5442 - val_accuracy: 0.3319 - val_loss: 1.8011
Epoch 2/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 43s 28ms/step - accuracy: 0.4167 - loss: 1.5948 - val_accuracy: 0.5119 - val_loss: 1.3412
Epoch 3/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 43s 27ms/step - accuracy: 0.5314 - loss: 1.3068 - val_accuracy: 0.5534 - val_loss: 1.2764
Epoch 4/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 44s 28ms/step - accuracy: 0.6217 - loss: 1.0749 - val_accuracy: 0.5973 - val_loss: 1.1745
Epoch 5/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 45s 29ms/step - accuracy: 0.6825 - loss: 0.9064 - val_accuracy: 0.6020 - val_loss: 1.2212
Epoch 6/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 46s 29ms/step - accuracy: 0.7306 - loss: 0.7720 - val_accuracy: 0.5982 - val_loss: 1.3409
Epoch 7/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 45s 29ms/step - accuracy: 0.7754 - loss: 0.6355 - val_accuracy: 0.6069 - val_loss: 1.3016
Epoch 8/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 45s 29ms/step - accuracy: 0.8175 -

accuracy: Akurasi pada data pelatihan.
loss: Nilai kehilangan pada data pelatihan.
val_accuracy: Akurasi pada data validasi.
val_loss: Nilai kehilangan pada data validasi.

Interpretasi Hasil: <br>
Peningkatan Akurasi pada Data Pelatihan
Akurasi pada data pelatihan meningkat dari 0.2018 pada epoch pertama hingga 0.8728 pada epoch ke-10.
Ini mengindikasikan bahwa model berhasil belajar dari data pelatihan dan menjadi semakin efektif dalam mengenali pola.

Penurunan Loss pada Data Pelatihan
Loss pada data pelatihan menurun dari 4.5442 pada epoch pertama menjadi 0.3716 pada epoch ke-10.
Penurunan loss menunjukkan bahwa model menjadi lebih baik dalam memprediksi label yang benar pada data pelatihan.

Stabilitas Akurasi pada Data Validasi
Akurasi pada data validasi cenderung tetap antara 0.3319 dan 0.6124, tidak menunjukkan tren peningkatan yang jelas.
Pada epoch ke-10, akurasi validasi adalah 0.5980, hampir sama dengan epoch sebelumnya.

Fluktuasi Loss pada Data Validasi
Nilai loss pada data validasi berfluktuasi, mulai dari 1.8011 pada epoch pertama hingga mencapai 1.7049 pada epoch ke-10.

# Evaluasi Model

In [15]:
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print("Loss:", test_loss)
print("Accuracy:", test_accuracy)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.5962 - loss: 1.6907
Loss: 1.7048835754394531
Accuracy: 0.5979999899864197


Akurasi pada data uji sekitar 59.8%. Ini berarti bahwa sekitar 60% prediksi model pada data uji adalah benar. Meskipun model telah belajar, akurasi ini menunjukkan ada ruang untuk peningkatan.

Loss pada data uji adalah 1.6907. Loss mengukur seberapa jauh prediksi model dari target sebenarnya. Nilai ini relatif tinggi, menunjukkan bahwa prediksi model masih jauh dari sempurna.

Akurasi pada data uji (59.8%) mirip dengan akurasi validasi yang terakhir (59.8%). Ini mengindikasikan bahwa model tidak mengalami perubahan drastis dalam performa antara data validasi dan data uji.

Nilai loss pada data uji (1.6907) juga konsisten dengan trend loss validasi pada akhir pelatihan (1.7049).